File: apps/backend/src/services/connection.service.ts
import { db } from '../config/database';
import { connection, Connection, NewConnection } from '../db/schema/connections';
import { users } from '../db/schema/users';
import { eq, and, desc } from 'drizzle-orm';
import { randomUUID } from 'crypto';
import { AppError } from '../utils/errors';

export const connectionService = {
  
  async createConnection(userId: string, data: {
    name: string;
    providerType: string;
    url: string;
    type?: string;
    authType?: string;
    authValue?: string;
    headers?: Record<string, string>;
    prefixId?: string;
    modelIds?: string[];
    tags?: string[];
    isDefault?: boolean;
  }): Promise<Connection> {
    const now = new Date();
    const id = randomUUID();

    
    if (data.isDefault) {
      await db.update(connection)
        .set({ isDefault: false, updatedAt: now })
        .where(eq(connection.userId, userId));
    }

    const [newConnection] = await db.insert(connection).values({
      id,
      userId,
      name: data.name,
      type: data.type || 'external',
      providerType: data.providerType,
      url: data.url,
      authType: data.authType || 'bearer',
      authValue: data.authValue,
      headers: data.headers ? JSON.stringify(data.headers) : null,
      prefixId: data.prefixId,
      modelIds: data.modelIds ? JSON.stringify(data.modelIds) : null,
      tags: data.tags ? JSON.stringify(data.tags) : null,
      isDefault: data.isDefault || false,
      isEnabled: true,
      priority: 0,
      createdAt: now,
      updatedAt: now,
    }).returning();

    return newConnection;
  },

  
  async getConnections(userId: string): Promise<Connection[]> {
    return db.select()
      .from(connection)
      .where(eq(connection.userId, userId))
      .orderBy(desc(connection.priority), desc(connection.createdAt));
  },

  
  async getConnectionById(id: string, userId: string): Promise<Connection> {
    const [conn] = await db.select()
      .from(connection)
      .where(and(eq(connection.id, id), eq(connection.userId, userId)));

    if (!conn) {
      throw new AppError(404, 'Connection not found');
    }

    return conn;
  },

  
  async getDefaultConnection(userId: string): Promise<Connection | null> {
    const [conn] = await db.select()
      .from(connection)
      .where(and(
        eq(connection.userId, userId),
        eq(connection.isDefault, true),
        eq(connection.isEnabled, true)
      ));

    return conn || null;
  },

  
  async getAnyEnabledConnection(userId: string): Promise<Connection | null> {
    const [conn] = await db.select()
      .from(connection)
      .where(and(
        eq(connection.userId, userId),
        eq(connection.isEnabled, true)
      ))
      .orderBy(desc(connection.priority))
      .limit(1);

    return conn || null;
  },

  
  async hasConnections(userId: string): Promise<boolean> {
    
    const userConnections = await db.select({ id: connection.id })
      .from(connection)
      .where(and(
        eq(connection.userId, userId),
        eq(connection.isEnabled, true)
      ))
      .limit(1);

    if (userConnections.length > 0) return true;

    
    const [master] = await db.select().from(users).where(eq(users.isMaster, true)).limit(1);
    if (master) {
      const masterConnections = await db.select({ id: connection.id })
        .from(connection)
        .where(and(
          eq(connection.userId, master.id),
          eq(connection.isEnabled, true)
        ))
        .limit(1);

      if (masterConnections.length > 0) return true;
    }

    return false;
  },

  
  async updateConnection(id: string, userId: string, data: Partial<{
    name: string;
    providerType: string;
    url: string;
    type: string;
    authType: string;
    authValue: string;
    headers: Record<string, string>;
    prefixId: string;
    modelIds: string[];
    tags: string[];
    isDefault: boolean;
    isEnabled: boolean;
    priority: number;
  }>): Promise<Connection> {
    const existing = await this.getConnectionById(id, userId);
    const now = new Date();

    
    if (data.isDefault) {
      await db.update(connection)
        .set({ isDefault: false, updatedAt: now })
        .where(and(eq(connection.userId, userId), eq(connection.isDefault, true)));
    }

    const updateData: any = { updatedAt: now };

    if (data.name !== undefined) updateData.name = data.name;
    if (data.providerType !== undefined) updateData.providerType = data.providerType;
    if (data.url !== undefined) updateData.url = data.url;
    if (data.type !== undefined) updateData.type = data.type;
    if (data.authType !== undefined) updateData.authType = data.authType;
    if (data.authValue !== undefined) updateData.authValue = data.authValue;
    if (data.headers !== undefined) updateData.headers = JSON.stringify(data.headers);
    if (data.prefixId !== undefined) updateData.prefixId = data.prefixId;
    if (data.modelIds !== undefined) updateData.modelIds = JSON.stringify(data.modelIds);
    if (data.tags !== undefined) updateData.tags = JSON.stringify(data.tags);
    if (data.isDefault !== undefined) updateData.isDefault = data.isDefault;
    if (data.isEnabled !== undefined) updateData.isEnabled = data.isEnabled;
    if (data.priority !== undefined) updateData.priority = data.priority;

    const [updated] = await db.update(connection)
      .set(updateData)
      .where(eq(connection.id, id))
      .returning();

    return updated;
  },

  
  async deleteConnection(id: string, userId: string): Promise<void> {
    await this.getConnectionById(id, userId);
    await db.delete(connection).where(eq(connection.id, id));
  },

  
  async verifyConnection(id: string, userId: string): Promise<{ success: boolean; message: string; models?: string[] }> {
    const conn = await this.getConnectionById(id, userId);

    try {
      const headers: Record<string, string> = {
        'Content-Type': 'application/json',
      };

      
      if (conn.authType === 'bearer' && conn.authValue) {
        headers['Authorization'] = `Bearer ${conn.authValue}`;
      } else if (conn.authType === 'api_key' && conn.authValue) {
        headers['X-API-Key'] = conn.authValue;
      }

      
      if (conn.headers) {
        const customHeaders = JSON.parse(conn.headers);
        Object.assign(headers, customHeaders);
      }

      
      const modelsUrl = conn.url.endsWith('/')
        ? `${conn.url}models`
        : `${conn.url}/models`;

      const response = await fetch(modelsUrl, {
        method: 'GET',
        headers,
      });

      if (!response.ok) {
        throw new Error(`HTTP ${response.status}: ${response.statusText}`);
      }

      const data = await response.json() as { data?: { id: string }[]; models?: { name: string }[] };
      const models = data.data?.map((m) => m.id) || data.models?.map((m) => m.name) || [];

      
      await db.update(connection)
        .set({ lastVerifiedAt: new Date(), updatedAt: new Date() })
        .where(eq(connection.id, id));

      return {
        success: true,
        message: 'Connection verified successfully',
        models,
      };
    } catch (error: any) {
      return {
        success: false,
        message: `Connection failed: ${error.message}`,
      };
    }
  },

  
  async getModelsFromConnection(id: string, userId: string): Promise<string[]> {
    const result = await this.verifyConnection(id, userId);
    return result.models || [];
  },

  
  async getAllModelsFromConnections(userId: string): Promise<{
    id: string;
    name: string;
    provider: string;
    providerType: string;
    connectionId: string;
    connectionName: string;
  }[]> {
    const connections = await db.select()
      .from(connection)
      .where(and(
        eq(connection.userId, userId),
        eq(connection.isEnabled, true)
      ))
      .orderBy(desc(connection.priority));

    const allModels: {
      id: string;
      name: string;
      provider: string;
      providerType: string;
      connectionId: string;
      connectionName: string;
    }[] = [];

    for (const conn of connections) {
      try {
        const headers: Record<string, string> = {
          'Content-Type': 'application/json',
        };

        if (conn.authType === 'bearer' && conn.authValue) {
          headers['Authorization'] = `Bearer ${conn.authValue}`;
        } else if (conn.authType === 'api_key' && conn.authValue) {
          headers['X-API-Key'] = conn.authValue;
        }

        if (conn.headers) {
          const customHeaders = JSON.parse(conn.headers);
          Object.assign(headers, customHeaders);
        }

        const modelsUrl = conn.url.endsWith('/')
          ? `${conn.url}models`
          : `${conn.url}/models`;

        const response = await fetch(modelsUrl, {
          method: 'GET',
          headers,
        });

        if (response.ok) {
          const data = await response.json() as { data?: { id?: string; name?: string }[]; models?: { id?: string; name?: string }[] };

          
          const models = data.data || data.models || [];

          for (const model of models) {
            const modelId = model.id || model.name || '';
            const modelName = model.name || model.id || '';

            if (!modelId) continue;

            
            if (conn.modelIds) {
              const allowedIds = JSON.parse(conn.modelIds);
              if (allowedIds.length > 0 && !allowedIds.includes(modelId)) {
                continue;
              }
            }

            
            const finalId = conn.prefixId ? `${conn.prefixId}${modelId}` : modelId;

            allModels.push({
              id: finalId,
              name: modelName,
              provider: conn.name,
              providerType: conn.providerType,
              connectionId: conn.id,
              connectionName: conn.name,
            });
          }
        }
      } catch (error) {
        
        console.error(`Failed to fetch models from ${conn.name}:`, error);
      }
    }

    return allModels;
  },
};

File: apps/backend-python/src/services/connection_service.py
import json
from datetime import datetime
from nanoid import generate
from sqlalchemy.orm import Session
from src.db.models import Connection
from src.utils.errors import NotFoundError
import httpx


def _serialize(conn: Connection, hide_auth: bool = False) -> dict:
    return {
        "id": conn.id,
        "userId": conn.user_id,
        "name": conn.name,
        "type": conn.type,
        "providerType": conn.provider_type,
        "url": conn.url,
        "authType": conn.auth_type,
        "authValue": "********" if hide_auth and conn.auth_value else conn.auth_value,
        "headers": json.loads(conn.headers) if isinstance(conn.headers, str) and conn.headers else None,
        "prefixId": conn.prefix_id,
        "modelIds": json.loads(conn.model_ids) if isinstance(conn.model_ids, str) and conn.model_ids else [],
        "tags": json.loads(conn.tags) if isinstance(conn.tags, str) and conn.tags else [],
        "isEnabled": conn.is_enabled,
        "isDefault": conn.is_default,
        "priority": conn.priority,
        "lastVerifiedAt": conn.last_verified_at.isoformat() if conn.last_verified_at else None,
        "createdAt": conn.created_at.isoformat() if conn.created_at else None,
        "updatedAt": conn.updated_at.isoformat() if conn.updated_at else None,
    }


class ConnectionService:
    def create_connection(self, db: Session, user_id: str, data: dict) -> dict:
        now = datetime.utcnow()
        conn = Connection(
            id=generate(), user_id=user_id,
            name=data["name"],
            type=data.get("type", "external"),
            provider_type=data["providerType"],
            url=data["url"],
            auth_type=data.get("authType", "bearer"),
            auth_value=data.get("authValue"),
            headers=json.dumps(data["headers"]) if data.get("headers") else None,
            prefix_id=data.get("prefixId"),
            model_ids=json.dumps(data.get("modelIds", [])),
            tags=json.dumps(data.get("tags", [])),
            is_enabled=True,
            is_default=data.get("isDefault", False),
            priority=data.get("priority", 0),
            created_at=now, updated_at=now,
        )
        db.add(conn)
        db.commit()
        db.refresh(conn)
        return _serialize(conn, hide_auth=True)

    def get_connections(self, db: Session, user_id: str) -> list:
        conns = db.query(Connection).filter(Connection.user_id == user_id).all()
        return [_serialize(c, hide_auth=True) for c in conns]

    def get_connection_by_id(self, db: Session, conn_id: str, user_id: str) -> dict:
        conn = db.query(Connection).filter(Connection.id == conn_id, Connection.user_id == user_id).first()
        if not conn:
            raise NotFoundError("Connection not found")
        return _serialize(conn, hide_auth=True)

    def get_connection_raw(self, db: Session, conn_id: str, user_id: str) -> Connection:
        conn = db.query(Connection).filter(Connection.id == conn_id, Connection.user_id == user_id).first()
        if not conn:
            raise NotFoundError("Connection not found")
        return conn

    def update_connection(self, db: Session, conn_id: str, user_id: str, data: dict) -> dict:
        conn = db.query(Connection).filter(Connection.id == conn_id, Connection.user_id == user_id).first()
        if not conn:
            raise NotFoundError("Connection not found")
        field_map = {
            "name": "name", "type": "type", "providerType": "provider_type",
            "url": "url", "authType": "auth_type", "authValue": "auth_value",
            "prefixId": "prefix_id", "isEnabled": "is_enabled",
            "isDefault": "is_default", "priority": "priority",
        }
        for key, attr in field_map.items():
            if key in data:
                setattr(conn, attr, data[key])
        if "headers" in data:
            conn.headers = json.dumps(data["headers"]) if data["headers"] else None
        if "modelIds" in data:
            conn.model_ids = json.dumps(data["modelIds"])
        if "tags" in data:
            conn.tags = json.dumps(data["tags"])
        conn.updated_at = datetime.utcnow()
        db.commit()
        db.refresh(conn)
        return _serialize(conn, hide_auth=True)

    def delete_connection(self, db: Session, conn_id: str, user_id: str) -> None:
        conn = db.query(Connection).filter(Connection.id == conn_id, Connection.user_id == user_id).first()
        if not conn:
            raise NotFoundError("Connection not found")
        db.delete(conn)
        db.commit()

    def verify_connection(self, db: Session, conn_id: str, user_id: str) -> dict:
        conn = db.query(Connection).filter(Connection.id == conn_id, Connection.user_id == user_id).first()
        if not conn:
            raise NotFoundError("Connection not found")
        try:
            headers = {"Content-Type": "application/json"}
            if conn.auth_type == "bearer" and conn.auth_value:
                headers["Authorization"] = f"Bearer {conn.auth_value}"
            elif conn.auth_type == "api_key" and conn.auth_value:
                headers["Authorization"] = f"Bearer {conn.auth_value}"
            resp = httpx.get(f"{conn.url.rstrip('/')}/v1/models", headers=headers, timeout=5)
            ok = resp.status_code < 400
            conn.last_verified_at = datetime.utcnow()
            db.commit()
            return {"success": ok, "status": resp.status_code}
        except Exception as e:
            return {"success": False, "error": str(e)}

    def get_models_from_connection(self, db: Session, conn_id: str, user_id: str) -> list:
        conn = db.query(Connection).filter(Connection.id == conn_id, Connection.user_id == user_id).first()
        if not conn:
            raise NotFoundError("Connection not found")
        try:
            headers = {}
            if conn.auth_type in ("bearer", "api_key") and conn.auth_value:
                headers["Authorization"] = f"Bearer {conn.auth_value}"
            resp = httpx.get(f"{conn.url.rstrip('/')}/v1/models", headers=headers, timeout=5)
            resp.raise_for_status()
            data = resp.json()
            return data.get("data", [])
        except Exception:
            return []

    def get_all_models_from_connections(self, db: Session, user_id: str) -> list:
        conns = db.query(Connection).filter(
            Connection.user_id == user_id, Connection.is_enabled == True
        ).all()
        models = []
        for conn in conns:
            try:
                headers = {}
                if conn.auth_type in ("bearer", "api_key") and conn.auth_value:
                    headers["Authorization"] = f"Bearer {conn.auth_value}"
                resp = httpx.get(f"{conn.url.rstrip('/')}/v1/models", headers=headers, timeout=5)
                resp.raise_for_status()
                data = resp.json()
                for m in data.get("data", []):
                    prefix = conn.prefix_id or ""
                    model_id = f"{prefix}{m.get('id', '')}" if prefix else m.get("id", "")
                    models.append({**m, "id": model_id, "connectionId": conn.id})
            except Exception:
                continue
        return models

    def has_connections(self, db: Session, user_id: str) -> bool:
        return db.query(Connection).filter(
            Connection.user_id == user_id, Connection.is_enabled == True
        ).count() > 0


connection_service = ConnectionService()
